src/Native/System.Security.Cryptography.Native/pal_ssl.cpp

```diff
@@ -390,3 +440,134 @@ extern "C" int32_t SslRead(SSL* ssl, void* buf, int32_t num)
 {
     return SSL_read(ssl, buf, num);
 }
+
+extern "C" int32_t IsSslRenegotiatePending(SSL* ssl)
+{
+    return SSL_renegotiate_pending(ssl) != 0;
+}
+
+extern "C" int32_t SslShutdown(SSL* ssl)
+{
+    return SSL_shutdown(ssl);
+}
+
+extern "C" void SslSetBio(SSL* ssl, BIO* rbio, BIO* wbio)
+{
+    SSL_set_bio(ssl, rbio, wbio);
+}
+
+extern "C" int32_t SslDoHandshake(SSL* ssl)
+{
+    return SSL_do_handshake(ssl);
+}
+
+extern "C" int32_t IsSslStateOK(SSL* ssl)
+{
+    return SSL_state(ssl) == SSL_ST_OK;
+}
+
+extern "C" X509* SslGetPeerCertificate(SSL* ssl)
+{
+    return SSL_get_peer_certificate(ssl);
+}
+
+extern "C" X509Stack* SslGetPeerCertChain(SSL* ssl)
+{
+    return SSL_get_peer_cert_chain(ssl);
+}
+
+extern "C" int32_t SslCtxUseCertificate(SSL_CTX* ctx, X509* x)
+{
+    return SSL_CTX_use_certificate(ctx, x);
+}
+
+extern "C" int32_t SslCtxUsePrivateKey(SSL_CTX* ctx, EVP_PKEY* pkey)
+{
+    return SSL_CTX_use_PrivateKey(ctx, pkey);
+}
+
+extern "C" int32_t SslCtxCheckPrivateKey(SSL_CTX* ctx)
+{
+    return SSL_CTX_check_private_key(ctx);
+}
+
+extern "C" int32_t BioCtrlPending(BIO* bio)
+{
+    size_t result = BIO_ctrl_pending(bio);
+    assert(result <= INT32_MAX);
+    return static_cast<int32_t>(result);
+}
+
+extern "C" void SslCtxSetQuietShutdown(SSL_CTX* ctx)
+{
+    SSL_CTX_set_quiet_shutdown(ctx, 1);
+}
+
+extern "C" X509NameStack* SslGetClientCAList(SSL* ssl)
+{
+    return SSL_get_client_CA_list(ssl);
+}
+
+extern "C" void SslCtxSetVerify(SSL_CTX* ctx, SslCtxSetVerifyCallback callback)
+{
+    int mode = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
+
+    SSL_CTX_set_verify(ctx, mode, callback);
+}
+
+// delimiter ":" is used to allow more than one strings
+// below string is corresponding to "AllowNoEncryption"
+#define SSL_TXT_Separator ":"
+#define SSL_TXT_AllIncludingNull SSL_TXT_ALL SSL_TXT_Separator SSL_TXT_eNULL
+
+extern "C" void SetEncryptionPolicy(SSL_CTX* ctx, EncryptionPolicy policy)
+{
+    const char* cipherString = nullptr;
+    switch (policy)
+    {
+        case EncryptionPolicy::RequireEncryption:
+            cipherString = SSL_TXT_ALL;
+            break;
+
+        case EncryptionPolicy::AllowNoEncryption:
+            cipherString = SSL_TXT_AllIncludingNull;
+            break;
+
+        case EncryptionPolicy::NoEncryption:
+            cipherString = SSL_TXT_eNULL;
+            break;
+    }
+
+    assert(cipherString != nullptr);
+
+    int result = SSL_CTX_set_cipher_list(ctx, cipherString);
+    assert(result == 1);
+}
+
+extern "C" void SslCtxSetClientCAList(SSL_CTX* ctx, X509NameStack* list)
+{
+    SSL_CTX_set_client_CA_list(ctx, list);
+}
+
+extern "C" void GetStreamSizes(int32_t* header, int32_t* trailer, int32_t* maximumMessage)
+{
+    if (header)
+    {
+        *header = SSL3_RT_HEADER_LENGTH;
+    }
+
+    if (trailer)
+    {
+        // TODO (Issue #3362) : Trailer size requirement is changing based on protocol
```

Comment:
I think the signature can (and probably will) change when this actually gets implemented. It won't be hard to change it then, and since I don't know what will be involved, it doesn't make sense to try and guess now.